### ISSUE TITLE
Pulumi uv removes packages in optional-groups

### ISSUE THREAD

**Issue by janrito** [pulumi/pulumi#17853]
### What happened?

In a uv envrionment, one can define optional dependencies. In fact, I'd like to define pulumi as an optional dependency, in group called `provisioning`. 

However, any uv command will resync the project, installing only the default dependencies. 

### Example

```[project]
name = "remote-analysis"
version = "0.1.0"
description = "Provision temporary instance to run some analysis"
readme = "README.md"
authors = [
    { name = "Alejandro Giacometti", email = "alejandro.giacometti@beauhurst.com" },
]
requires-python = ">=3.12"
dependencies = [
    "duckdb-engine>=0.13.5",
    "duckdb>=1.1.3",
    "ipykernel>=6.29.5",
    "jupysql>=0.10.16",
    "jupyterlab>=4.3.1",
    "polars>=1.14.0",
    "psycopg2-binary>=2.9.10",
    "pyarrow>=18.0.0",
    "sqlalchemy>=2.0.36",
]

[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[dependency-groups]
dev = ["ruff>=0.7.4", "sqlfluff>=3.2.5"]
provision = [
    "pulumi-command>=1.0.1",
    "pulumi>=3.138.0",
    "pulumi-aws>=6.59.1",
]
```

### Output of `pulumi about`

```CLI
Version      3.141.0
Go Version   go1.23.3
Go Compiler  gc

Plugins
KIND      NAME    VERSION
language  python  3.141.0

Host
OS       darwin
Version  14.4.1
Arch     arm64

```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by julienp** [comment]
You can add the provision group to the default groups

```
[tool.uv]
default-groups = ["provision"]
```

This way `uv sync` will include the group. Does that solve your problem or do you need more flexibility?

**Comment by janrito** [comment]
Thanks, this does indeed keep the `provision` group installed. But it will still remove packages in other groups. 

I guess that there is the option of not using pulumi to manage my environment, but that is also currently broken due to this https://github.com/pulumi/pulumi/issues/12390

**Comment by julienp** [comment]
`default-group` could list all your groups to prevent the other dependencies from being removed. What would be the reason to have groups then though, other than some organisation of dependencies?

Right now `pulumi install` will run `uv sync`. We could make it run `uv sync --group $some_groups`  with a `group` option in `Pulumi.yaml`. That would still remove other packages not in the group from the venv though.

What is the reason for having a separate `provision` group in the first place? Do you have downstream users that install your package and you don't want them to have the pulumi package installed by default? What other groups do you have in your pyproject, and what's their purpose? I am trying to understand the use case, so we can come up with the best solution here.

**Comment by janrito** [comment]
Groups can exist for a huge variety of reasons.

My usecase the provisioning of the application is something that happens rarely, I want provisioning dependencies only to be installed explicitly and when needed. Other similar dependency groups are `dev`, `test` and `debug`

**Comment by julienp** [comment]
I think adding a new `groups` option to Pulumi.yaml as outlined above will help here, but it would still be useful to better understand how you expect users to use the other groups.

If there are groups `provision` and `debug`, and `Pulumi.yaml` only sets the extra group `provision`, then `pulumi install` will still drop the deps from `debug`. And if users run `uv sync --group debug`, it will remove the deps from `provision`.

**Comment by janrito** [comment]
Yeah, i think the issue here is that uv does not currently have a way to remember which groups or extras have been installed: https://github.com/astral-sh/uv/issues/6001

Users will use groups for many purposes. The one that I've been thinking is that I would like to put pulumi's requirements in an isolated group that is not installed by default. 

The ideal situation would be that pulumi does not change the state of the virtual environment when it is run
